File: python/src/fluke_3540/snapshots.py
"""Pick "normal operation" snapshots — quiet windows useful as baseline plots.

A snapshot is a 5-minute window where total active power is at its calmest
(low rolling stdev) and which does not overlap any detected event. Useful for
showing what the system looks like *between* anomalies.

The rolling-stdev pass is O(N) via prefix sums of P and P², so it scales to
week-long sessions. ``pick_snapshots`` accepts either a
:class:`~fluke_3540.store.ColumnStore` (the production path) or an iterable of
:class:`~fluke_3540.parser.Record` (small fixtures / legacy callers).
"""
from __future__ import annotations

import datetime as dt
import math
from dataclasses import dataclass
from typing import Iterable, Sequence

from .events import Event
from .store import ColumnStore


@dataclass(frozen=True)
class Snapshot:
    id: int
    t_start: dt.datetime
    t_end: dt.datetime
    t_center: dt.datetime
    p_total_mean_w: float
    p_total_stdev_w: float


def _prefix_sums(values: Sequence[float]) -> tuple[list[float], list[float]]:
    """Return (prefix_sum, prefix_sum_of_squares) with a leading 0.

    ``ps[k]`` is the sum of values[:k]; ``ps2[k]`` the sum of squares of
    values[:k]. The sum over [lo, hi) is ``ps[hi] - ps[lo]``.
    """
    n = len(values)
    ps = [0.0] * (n + 1)
    ps2 = [0.0] * (n + 1)
    s = 0.0
    s2 = 0.0
    for i, v in enumerate(values):
        s += v
        s2 += v * v
        ps[i + 1] = s
        ps2[i + 1] = s2
    return ps, ps2


def _window_mean_stdev(ps: list[float], ps2: list[float],
                       lo: int, hi: int) -> tuple[float, float]:
    """Population mean and stdev over values[lo:hi] from prefix sums."""
    count = hi - lo
    if count <= 0:
        return 0.0, 0.0
    total = ps[hi] - ps[lo]
    total_sq = ps2[hi] - ps2[lo]
    mean = total / count
    var = total_sq / count - mean * mean
    if var < 0.0:  # tiny negatives from float cancellation
        var = 0.0
    return mean, math.sqrt(var)


def _rolling_stdev(values: Sequence[float], window: int) -> list[float | None]:
    """Right-aligned rolling population stdev via prefix sums (O(N)).

    First ``window-1`` entries are None, matching the original semantics.
    """
    out: list[float | None] = [None] * len(values)
    n = len(values)
    if window <= 1 or window > n:
        return out
    ps, ps2 = _prefix_sums(values)
    for i in range(window - 1, n):
        _, sd = _window_mean_stdev(ps, ps2, i - window + 1, i + 1)
        out[i] = sd
    return out


def _as_store(records_or_store) -> ColumnStore:
    if isinstance(records_or_store, ColumnStore):
        return records_or_store
    return ColumnStore.from_records(records_or_store)


def pick_snapshots(
    records_or_store: Iterable | ColumnStore,
    events: Iterable[Event],
    n: int = 3,
    window_secs: int = 300,
    min_separation_secs: int = 3600,
) -> list[Snapshot]:
    """Pick up to N snapshots: low-stdev windows that don't overlap events.

    Snapshots span ``window_secs`` records ending at the chosen index.
    Consecutive picks are separated by ≥ ``min_separation_secs``.

    Accepts either a ColumnStore or an iterable of Records.
    """
    store = _as_store(records_or_store)
    nrec = store.n
    if nrec == 0:
        return []
    p = store.col("P_total_avg_W")
    rolling = _rolling_stdev(p, window_secs)
    times = list(store.iter_times())
    end_times = list(store.iter_end_times())

    ps, _ps2 = _prefix_sums(p)  # for window means

    event_intervals = [(ev.t_start, ev.t_end) for ev in events]

    def overlaps_event(i: int) -> bool:
        if rolling[i] is None:
            return True
        win_start = times[i - window_secs + 1]
        win_end = end_times[i]
        for es, ee in event_intervals:
            if not (win_end < es or win_start > ee):
                return True
        return False

    candidates = [
        (rolling[i], i) for i in range(nrec)
        if rolling[i] is not None and not overlaps_event(i)
    ]
    candidates.sort(key=lambda x: x[0])  # lowest stdev first

    picked: list[Snapshot] = []
    used_times: list[dt.datetime] = []
    for stdev_val, i in candidates:
        win_start = times[i - window_secs + 1]
        win_end = end_times[i]
        center = win_start + (win_end - win_start) / 2
        if any(
            abs((center - prev).total_seconds()) < min_separation_secs
            for prev in used_times
        ):
            continue
        lo = i - window_secs + 1
        win_mean = (ps[i + 1] - ps[lo]) / window_secs
        picked.append(Snapshot(
            id=len(picked),
            t_start=win_start, t_end=win_end, t_center=center,
            p_total_mean_w=win_mean,
            p_total_stdev_w=stdev_val,
        ))
        used_times.append(center)
        if len(picked) >= n:
            break

    return picked

File: web/snapshots.js
// Snapshot picking — port of python/src/fluke_3540/snapshots.py.
// Picks quiet, non-event windows by rolling stdev of P_total.

import { asColumnSource } from './column_source.js';

function pstdev(values, start, end) {
  if (end - start + 1 < 2) return 0;
  let s = 0;
  for (let i = start; i <= end; i++) s += values[i];
  const m = s / (end - start + 1);
  let sq = 0;
  for (let i = start; i <= end; i++) sq += (values[i] - m) * (values[i] - m);
  return Math.sqrt(sq / (end - start + 1));
}

function mean(values, start, end) {
  let s = 0;
  for (let i = start; i <= end; i++) s += values[i];
  return s / (end - start + 1);
}

/**
 * Pick up to N snapshot windows that are quiet (low P_total stdev) and don't
 * overlap any detected event.
 * @param {Array<{startMs:number, endMs:number, floats:Float32Array}>|import('./column_store.js').ColumnStore} source
 * @param {Array<{tStartMs:number, tEndMs:number}>} events
 * @param {object} spec
 * @param {{n?: number, windowSecs?: number, minSeparationSecs?: number}} [opts]
 */
export function pickSnapshots(source, events, spec, opts = {}) {
  const { n = 3, windowSecs = 300, minSeparationSecs = 3600 } = opts;
  const src = asColumnSource(source, spec);
  if (src.length === 0) return [];

  const p = src.column('P_total_avg_W');
  const N = src.length;

  if (windowSecs <= 1 || windowSecs > N) return [];

  // Rolling pstdev, right-aligned. Cheap to compute incrementally but the
  // straightforward window pass is fast enough for 75 K records.
  const rolling = new Array(N).fill(null);
  for (let i = windowSecs - 1; i < N; i++) {
    rolling[i] = pstdev(p, i - windowSecs + 1, i);
  }

  const eventIntervals = events.map((ev) => [ev.tStartMs, ev.tEndMs]);
  function overlapsEvent(i) {
    if (rolling[i] === null) return true;
    const winStart = src.startMs(i - windowSecs + 1);
    const winEnd = src.endMs(i);
    for (const [es, ee] of eventIntervals) {
      if (!(winEnd < es || winStart > ee)) return true;
    }
    return false;
  }

  const candidates = [];
  for (let i = 0; i < N; i++) {
    if (rolling[i] !== null && !overlapsEvent(i)) {
      candidates.push([rolling[i], i]);
    }
  }
  candidates.sort((a, b) => a[0] - b[0]);

  const picked = [];
  const usedCenters = [];
  for (const [stdevVal, i] of candidates) {
    const winStart = src.startMs(i - windowSecs + 1);
    const winEnd = src.endMs(i);
    const center = (winStart + winEnd) / 2;
    let tooClose = false;
    for (const prev of usedCenters) {
      if (Math.abs(center - prev) < minSeparationSecs * 1000) { tooClose = true; break; }
    }
    if (tooClose) continue;
    const meanP = mean(p, i - windowSecs + 1, i);
    picked.push({
      id: picked.length,
      tStartMs: winStart,
      tEndMs: winEnd,
      tCenterMs: center,
      pTotalMeanW: meanP,
      pTotalStdevW: stdevVal,
    });
    usedCenters.push(center);
    if (picked.length >= n) break;
  }
  return picked;
}
